ir/program/block.rs

```diff
@@ -106,37 +119,171 @@ impl FunctionalBlockBuilder {
 }
 
 #[derive(Debug, Clone)]
-pub struct BlockContext {
+pub struct VariableRegistry {
     variable_names: HashMap<Variable, String>,
-    variable_declaration: HashMap<Variable, ScopeId>,
-    variable_names_index: HashMap<String, Variable>,
     variable_id_allocator: u16,
-
-    scope_id_allocator: u16,
-    scope_parents: HashMap<ScopeId, ScopeId>,
-
     variable_categories: HashMap<Variable, (VariableCategory, Constraint<Variable>)>,
     variable_optionality: HashMap<Variable, VariableOptionality>,
 }
 
-impl BlockContext {
-    pub fn new() -> BlockContext {
+impl VariableRegistry {
+    pub(crate) fn new() -> VariableRegistry {
         Self {
             variable_names: HashMap::new(),
-            variable_declaration: HashMap::new(),
-            variable_names_index: HashMap::new(),
             variable_id_allocator: 0,
-            scope_id_allocator: 1, // `0` is reserved for ROOT
-            scope_parents: HashMap::new(),
             variable_categories: HashMap::new(),
             variable_optionality: HashMap::new(),
         }
     }
 
```

Comment:
BlockContext has been split to have a reference to a pipeline level `VariableRegistry`